File: packages/corridor/src/geo.ts
import type { LatLng } from "./types.ts";

/** IUGG mean Earth radius, metres. */
const R = 6371008.8;
const DEG = Math.PI / 180;

export function haversineM(a: LatLng, b: LatLng): number {
  const dLat = (b.lat - a.lat) * DEG;
  const dLng = (b.lng - a.lng) * DEG;
  const s =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(a.lat * DEG) * Math.cos(b.lat * DEG) * Math.sin(dLng / 2) ** 2;
  return 2 * R * Math.asin(Math.min(1, Math.sqrt(s)));
}

type Local = { x: number; y: number };

/**
 * Equirectangular projection about `origin`. Over a corridor tens of kilometres
 * across at Jordanian latitudes the error is well under a metre, which is far
 * finer than any distance this system reports.
 */
function toLocal(p: LatLng, origin: LatLng): Local {
  const cosLat = Math.cos(origin.lat * DEG);
  return {
    x: (p.lng - origin.lng) * DEG * R * cosLat,
    y: (p.lat - origin.lat) * DEG * R,
  };
}

/** A reference path with cumulative distances precomputed. */
export type PathIndex = {
  points: LatLng[];
  local: Local[];
  /** cumulativeM[i] is the distance from the path start to points[i]. */
  cumulativeM: number[];
  totalM: number;
  origin: LatLng;
};

export function indexPath(points: LatLng[]): PathIndex {
  if (points.length < 2) throw new Error("a path needs at least two points");
  const origin = points[0];
  const local = points.map((p) => toLocal(p, origin));
  const cumulativeM = [0];
  for (let i = 1; i < points.length; i++) {
    cumulativeM.push(cumulativeM[i - 1] + haversineM(points[i - 1], points[i]));
  }
  return {
    points,
    local,
    cumulativeM,
    totalM: cumulativeM[cumulativeM.length - 1],
    origin,
  };
}

export type PathFix = {
  /** Distance from the path start to the projected point, in metres. */
  alongM: number;
  /** Perpendicular distance from the query point to the path, in metres. */
  offsetM: number;
};

/** Projects a point onto the nearest segment of a path. */
export function nearestOnPath(p: LatLng, idx: PathIndex): PathFix {
  const q = toLocal(p, idx.origin);
  let best: PathFix = { alongM: 0, offsetM: Infinity };

  for (let i = 0; i < idx.local.length - 1; i++) {
    const a = idx.local[i];
    const b = idx.local[i + 1];
    const dx = b.x - a.x;
    const dy = b.y - a.y;
    const len2 = dx * dx + dy * dy;

    // t is the clamped position along the segment, 0 at a and 1 at b.
    const t =
      len2 === 0 ? 0 : Math.max(0, Math.min(1, ((q.x - a.x) * dx + (q.y - a.y) * dy) / len2));

    const px = a.x + t * dx;
    const py = a.y + t * dy;
    const offsetM = Math.hypot(q.x - px, q.y - py);

    if (offsetM < best.offsetM) {
      const segmentM = idx.cumulativeM[i + 1] - idx.cumulativeM[i];
      best = { alongM: idx.cumulativeM[i] + t * segmentM, offsetM };
    }
  }
  return best;
}

File: packages/core/lib/src/geo.dart
import 'dart:math' as math;

/// A WGS84 coordinate.
///
/// These exist only on this side of the wire. Nothing in [transport] carries
/// one: the phone does its own geometry and sends how far along a line it is.
class LatLng {
  const LatLng(this.lat, this.lng);

  final double lat;
  final double lng;

  @override
  String toString() => 'LatLng($lat, $lng)';
}

const double _earthRadiusM = 6371008.8;
const double _deg = math.pi / 180;

double haversineM(LatLng a, LatLng b) {
  final dLat = (b.lat - a.lat) * _deg;
  final dLng = (b.lng - a.lng) * _deg;
  final s = math.pow(math.sin(dLat / 2), 2) +
      math.cos(a.lat * _deg) * math.cos(b.lat * _deg) * math.pow(math.sin(dLng / 2), 2);
  return 2 * _earthRadiusM * math.asin(math.min(1, math.sqrt(s.toDouble())));
}

class _Local {
  const _Local(this.x, this.y);
  final double x;
  final double y;
}

/// Equirectangular projection about [origin]. Over a corridor tens of
/// kilometres across at Jordanian latitudes the error is well under a metre,
/// far finer than any distance this system reports.
_Local _toLocal(LatLng p, LatLng origin) {
  final cosLat = math.cos(origin.lat * _deg);
  return _Local(
    (p.lng - origin.lng) * _deg * _earthRadiusM * cosLat,
    (p.lat - origin.lat) * _deg * _earthRadiusM,
  );
}

/// Where a point falls relative to a path.
class PathFix {
  const PathFix({required this.alongM, required this.offsetM});

  /// Distance from the path start to the projected point.
  final double alongM;

  /// Perpendicular distance from the query point to the path.
  final double offsetM;
}

/// A reference path with cumulative distances precomputed once.
class PathIndex {
  PathIndex(this.points)
      : assert(points.length >= 2, 'a path needs at least two points'),
        _origin = points.first,
        _local = <_Local>[],
        cumulativeM = <double>[0] {
    for (final p in points) {
      _local.add(_toLocal(p, _origin));
    }
    for (var i = 1; i < points.length; i++) {
      cumulativeM.add(cumulativeM[i - 1] + haversineM(points[i - 1], points[i]));
    }
  }

  final List<LatLng> points;
  final LatLng _origin;
  final List<_Local> _local;
  final List<double> cumulativeM;

  double get totalM => cumulativeM.last;

  /// Projects a point onto the nearest segment of this path.
  PathFix nearest(LatLng p) {
    final q = _toLocal(p, _origin);
    var bestAlong = 0.0;
    var bestOffset = double.infinity;

    for (var i = 0; i < _local.length - 1; i++) {
      final a = _local[i];
      final b = _local[i + 1];
      final dx = b.x - a.x;
      final dy = b.y - a.y;
      final len2 = dx * dx + dy * dy;

      // Clamped position along the segment: 0 at a, 1 at b.
      final t = len2 == 0
          ? 0.0
          : math.max(0.0, math.min(1.0, ((q.x - a.x) * dx + (q.y - a.y) * dy) / len2));

      final px = a.x + t * dx;
      final py = a.y + t * dy;
      final offset = math.sqrt(math.pow(q.x - px, 2) + math.pow(q.y - py, 2));

      if (offset < bestOffset) {
        bestOffset = offset;
        bestAlong = cumulativeM[i] + t * (cumulativeM[i + 1] - cumulativeM[i]);
      }
    }
    return PathFix(alongM: bestAlong, offsetM: bestOffset);
  }
}
